packages/ui/react-ui-form/src/components/Form/ArrayField.tsx

```diff
@@ -0,0 +1,103 @@
+//
+// Copyright 2025 DXOS.org
+//
+
+import { pipe } from 'effect';
+import { capitalize } from 'effect/String';
+import React from 'react';
+
+import { AST } from '@dxos/echo-schema';
+import { findNode, getDiscriminatedType, isDiscriminatedUnion, SimpleType } from '@dxos/effect';
+import { invariant } from '@dxos/invariant';
+import { IconButton, useTranslation } from '@dxos/react-ui';
+import { mx } from '@dxos/react-ui-theme';
+import { getSchemaProperties, type SchemaProperty } from '@dxos/schema';
+
+import { type ComponentLookup } from './Form';
+import { FormField } from './FormContent';
+import { useFormValues, type FormInputStateProps } from './FormContext';
+import { InputHeader, type InputComponent } from './Input';
+import { translationKey } from '../../translations';
+
+const padding = 'px-2';
+
+type ArrayFieldProps = {
+  property: SchemaProperty<any>;
+  readonly?: boolean;
+  inputProps: FormInputStateProps;
+  path?: (string | number)[];
+  Custom?: Partial<Record<string, InputComponent>>;
+  lookupComponent?: ComponentLookup;
+};
+
+export const ArrayField = ({ property, readonly, path, inputProps, Custom, lookupComponent }: ArrayFieldProps) => {
+  const { t } = useTranslation(translationKey);
+  const { ast, name, type, title } = property;
+  const values = useFormValues(path ?? []) as any[];
+  invariant(Array.isArray(values), 'Expected array values');
```

Comment:
no need for description for obvious non-semantic invariants, that are really just type checks.